qiskit/transpiler/passes/scheduling/align_measures.py

```diff
@@ -0,0 +1,127 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Align measurement instructions."""
+
+from collections import defaultdict
+from typing import List
+
+from qiskit.circuit.delay import Delay
+from qiskit.circuit.measure import Measure
+from qiskit.circuit.parameterexpression import ParameterExpression
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler import InstructionDurations
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+from qiskit.transpiler.passes.scheduling.alap import ALAPSchedule
+
+
+class AlignMeasures(TransformationPass):
+    """Measurement alignment.
+
+    This is control electronics aware optimization pass.
+    """
+
+    def __init__(self, alignment: int, durations: InstructionDurations):
+        super().__init__()
+        self.alignment = alignment
+        self.durations = durations
+
+    def run(self, dag: DAGCircuit):
+        """Run the measurement alignment pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): DAG to be checked.
+
+        Returns:
+            DAGCircuit: DAG with consistent timing and op nodes annotated with duration.
+
+        Raises:
+            TranspilerError: When instruction duration is not provided.
+        """
+        time_unit = self.property_set["time_unit"]
+
+        if len(dag.op_nodes(Delay)) == 0 or len(dag.op_nodes(Measure)) == 0:
+            # delay is only instruction that induce misalignment
+            # other instructions cannot be arbitrarily stretched
+            # if we don't find any delay skip rescheduling for performance reason
+            return dag
+
+        # if circuit is not yet scheduled, schedule with ALAP method
+        if dag.duration is None:
+            scheduler = ALAPSchedule(durations=self.durations)
+            scheduler.property_set["time_unit"] = time_unit
+            # run ALAP schedule to find measurement times
+            for required_pass in scheduler.requires:
+                dag = required_pass.run(dag)
+            dag = scheduler.run(dag)
```

Comment:
Thanks. I took the approach 1 with slight improvement of the logic. Because we don't want to reduce breaking change as much as possible, so new logic has validation of delay duration to trigger pass execution. If all delays are multiple of 16 this pass is not executed, thus calling transpile without scheduling option doesn't cause an error.

Reply:
I know, in general, we should not call another pass internally. I just want to note that we are accepting to break some existing codes using delays without scheduling, e.g. T1/T2 experiments, if requiring the external call of ALAP pass. They can be fixed only by supplying `scheduling_method` to `transpile` though. I'm wondering if this cost is allowable. What do you think? @ajavadia

Reply:
I think the existence of `alignment` should tell the transpiler to do scheduling, even if `scheduling_method` is not supplied. There is a similar issue here https://github.com/Qiskit/qiskit-terra/issues/6667

Reply:
> but they only declare that a certain pass must have been run before them, not that a certain property must have been set. I think @1ucian0 has been meaning to implement this.

it is not possible to do "require a property" right now. However, you can conditionally run a pass based on the existence of a property which, if I understand this correctly, is the case in this situation.